libff/algebra/fields/binary_field.hpp

```diff
@@ -0,0 +1,101 @@
+/** @file
+ *****************************************************************************
+ Declaration of common API for all finite fields in the binary/ directory.
+
+ Currently NOT used by the fields in this library. This class is not actually
+ the parent class of any field. All APIs are enforced through tests instead.
+
+ The reason for this is to ensure high performance of all fields. This class
+ exists as documentation for common API between fields.
+
+ Includes fields F2^n for specified n. All of the binary entension fields must
+ implement all functions declared in this class.
+ *****************************************************************************
+ * @author     This file is part of libff, developed by SCIPR Lab
+ *             and contributors (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+#include <libff/algebra/field_utils/bigint.hpp>
+#include <vector>
+
+namespace libff {
+
+template<typename T>
+class BinaryField;
+
+/* The type parameter T is intended to be set to the child class
+ * when this class is extended. For example,
+ * class gf32 : public BinaryField<gf32> ...
+ */
+template<typename T>
+class BinaryField {
+public:
+    /* Functions unique to binary fields */
+
+    /** Returns the constituent bits in 64 bit words, in little-endian order. */
+    std::vector<uint64_t> as_words() const = 0;
```

Comment:
Regarding from_words(), this would probably not be too hard, but why would it be needed? There are already the constructor, which takes in uint64_t as arguments, and the >> operator.

Reply:
For adding this to Fp^n, how would that be done? Fp has as_bigint() and as_ulong(), but it doesn't make as much sense to convert to a vector<uint64_t> for Fp or Fp^n. If it is needed to store the value in a string there is << and >>.

Reply:
In libiop we need to construct field elements from raw byte strings. Previously we did this by using the data field directly, which was a hack. I guess we could support it by implementing a from_bytes method?

From_words for Fp isn't so weird, there is a notion of words within bigint. (In particular, the bigint.data field)

It is a bit weird in Fp^n, but we could repeat the idea for Fp, `n` times.